lib/casbin/enforcer.rb

```diff
@@ -110,15 +104,20 @@ def has_permission_for_user(user, *permission)
     # g, role:admin, role:user
     # get_roles_for_user("alice") can only get: ["role:admin"].
     # But get_implicit_roles_for_user("alice") will get: ["role:admin", "role:user"].
-    def get_implicit_roles_for_user(name, *domain)
-      # stub method
-      puts "call get_implicit_roles_for_user with name = #{name} domain = #{domain}"
-    end
-
-    # gets the roles that a user has inside a domain.
-    def get_roles_for_user_in_domain(name, domain)
-      # stub method
-      puts "call get_roles_for_user_in_domain with name = #{name} domain = #{domain}"
+    def get_implicit_roles_for_user(name, domain = nil)
+      res = []
+      queue = Array.new([name])
+      while queue.size.positive?
+        name = queue.delete_at(0)
```

Comment:
Should we add some guard from the endless cycle?

Reply:
we cannot know the nesting depth